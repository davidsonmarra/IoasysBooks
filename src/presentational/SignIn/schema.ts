import * as Yup from 'yup';

export const schema = Yup.object().shape({
  email: Yup.string().email('Formato de e-mail inválido.').required('Campo Obrigatório.'),
  password: Yup.string().required('Campo Obrigatório.')
});
