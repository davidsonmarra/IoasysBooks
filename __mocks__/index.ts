import BookDTO from '../src/@types/BookDTO';

interface IMockFetchDataBooks {
  data: BookDTO[];
  page: number;
  totalPages: number;
  totalItems: number;
}

export const mockBook: BookDTO = {
  id: '1',
  title: 'Test Title',
  imageUrl: 'https://www.test.com',
  authors: ['Test Author'],
  pageCount: 100,
  publisher: 'Test Publisher',
  published: 1980,
  category: 'Test Category',
  language: 'Test Language',
  isbn10: 'Test ISBN10',
  isbn13: 'Test ISBN13',
  description: 'Test Description'
};

export const mockPreloadedStateEmpty = {
  profile: {
    email: '',
    token: '',
    error: {} as Error,
    isLogged: false,
    isLoading: false
  },
  books: {
    isEnd: false,
    loadingFetchBooks: false,
    error: {} as Error,
    booksData: [],
    category: { key: '', title: '' },
    search: ''
  }
};

export const mockPreloadedStateFilled = {
  profile: {
    email: 'davidson',
    token: 'teste',
    error: {} as Error,
    isLogged: true,
    isLoading: false
  },
  books: {
    isEnd: false,
    loadingFetchBooks: false,
    error: {} as Error,
    booksData: [mockBook],
    category: { key: '', title: '' },
    search: ''
  }
};

export const mockFetchDataBooks: IMockFetchDataBooks = {
  data: [
    {
      ...mockBook,
      id: '2',
      title: 'Harry Potter 1'
    },
    {
      ...mockBook,
      id: '3',
      title: 'Harry Potter 2'
    },
    {
      ...mockBook,
      id: '4',
      title: 'Harry Potter 3'
    },
    {
      ...mockBook,
      id: '5',
      title: 'Harry Potter 4'
    },
    {
      ...mockBook,
      id: '6',
      title: 'Harry Potter 5'
    }
  ],
  page: 1,
  totalPages: 1,
  totalItems: 2
};
