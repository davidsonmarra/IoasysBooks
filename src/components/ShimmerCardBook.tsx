import React from 'react';
import styled from 'styled-components/native';
import { RFValue } from 'react-native-responsive-fontsize';
import Animated, { FadeIn } from 'react-native-reanimated';

export function ShimmerCardBook() {
  return (
    <AnimatedWrapper entering={FadeIn}>
      <StyledContainer>
        <StyledImageBook />
        <StyledInfo>
          <StyledHeader>
            <StyledTitle />
            <StyledAuthor />
          </StyledHeader>
          <StyledFooter>
            <StyledInfoDetails />
            <StyledInfoDetails />
            <StyledInfoDetails />
          </StyledFooter>
        </StyledInfo>
      </StyledContainer>
    </AnimatedWrapper>
  );
}

const StyledWrapper = styled.View`
  width: 100%;
  padding: 0 ${RFValue(16)}px;
`;

const AnimatedWrapper = Animated.createAnimatedComponent(StyledWrapper);

const StyledContainer = styled.View`
  width: 100%;
  height: ${RFValue(180)}px;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  margin-bottom: ${RFValue(16)}px;
  padding: ${RFValue(8)}px ${RFValue(16)}px;
  border-radius: ${RFValue(4)}px;
  box-shadow: 0px 6px 24px rgba(84, 16, 95, 0.13);
  background-color: ${({ theme }) => theme.colors.shape};
`;

const StyledImageBook = styled.Image`
  width: ${RFValue(81)}px;
  height: ${RFValue(122)}px;
  margin-right: ${RFValue(16)}px;
  border-radius: ${RFValue(2)}px;
  background-color: ${({ theme }) => theme.colors.shimmer};
`;

const StyledInfo = styled.View`
  flex: 1;
  height: ${RFValue(180)}px;
  justify-content: space-between;
  padding: ${RFValue(16)}px 0;
`;

const StyledHeader = styled.View``;

const StyledTitle = styled.View`
  width: 60%;
  height: ${RFValue(12)}px;
  border-radius: ${RFValue(2)}px;
  margin-bottom: ${RFValue(2)}px;
  background-color: ${({ theme }) => theme.colors.shimmer};
`;

const StyledAuthor = styled.View`
  width: 45%;
  height: ${RFValue(12)}px;
  border-radius: ${RFValue(2)}px;
  background-color: ${({ theme }) => theme.colors.shimmer};
`;

const StyledFooter = styled.View``;

const StyledInfoDetails = styled.View`
  width: 45%;
  height: ${RFValue(12)}px;
  border-radius: ${RFValue(2)}px;
  margin-bottom: ${RFValue(4)}px;
  background-color: ${({ theme }) => theme.colors.shimmer};
`;
