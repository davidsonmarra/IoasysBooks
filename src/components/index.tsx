export { Input } from './form/Input';
export { SearchInput } from './form/SearchInput';
export { FilterModal } from './FilterModal';
export { CardBook } from './CardBook';
export { ShimmerCardBook } from './ShimmerCardBook';
