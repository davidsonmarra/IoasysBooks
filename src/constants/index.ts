const constants = {
  BASE_URL: 'https://books.ioasys.com.br/api/v1',
  TIMEOUT: 8000,
  asyncStorageUserKey: '@books_asyncStorageUserKey',
  asyncStorageUserId: '@books_asyncStorageUserId',
  asyncStorageUserRefresh: '@books_asyncStorageUserRefresh'
};

export default constants;
