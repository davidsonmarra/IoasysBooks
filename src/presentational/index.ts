export { SignIn } from './SignIn';
export { BookDetails } from './BookDetails';
export { Home } from './Home';
