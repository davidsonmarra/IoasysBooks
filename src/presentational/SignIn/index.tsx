import React, { useEffect } from 'react';
import {
  Keyboard,
  KeyboardAvoidingView,
  Platform,
  StatusBar,
  TouchableWithoutFeedback
} from 'react-native';
import styled from 'styled-components/native';
import { yupResolver } from '@hookform/resolvers/yup';
import { useForm } from 'react-hook-form';
import { RFValue } from 'react-native-responsive-fontsize';
import { schema } from './schema';
import { Input } from '../../components';
import TitleSvg from '../../assets/title.svg';
import LogoSvg from '../../assets/logo.svg';
import background from '../../assets/background.png';
import Toast from 'react-native-toast-message';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  interpolate,
  withTiming,
  Easing,
  Extrapolate
} from 'react-native-reanimated';
import { useDispatch, useSelector } from 'react-redux';
import { LOGIN } from '../../store/slices/profileSlice';
import { IRootState } from '../../store/store';
import { isObjectEmpty } from '../../utils';

export interface IFormSignIn {
  email: string;
  password: string;
}

export function SignIn() {
  const { error } = useSelector(({ profile }: IRootState) => profile);

  const dispatch = useDispatch();
  const animation = useSharedValue(0);

  const animationStyleLogo = useAnimatedStyle(() => ({
    marginRight: interpolate(animation.value, [0, 1], [-16.6, 16.6], Extrapolate.CLAMP),
    left: interpolate(animation.value, [0, 1], [50, 0], Extrapolate.CLAMP),
    opacity: interpolate(animation.value, [0, 1], [0, 1])
  }));

  const animationStyleTitle = useAnimatedStyle(() => ({
    right: interpolate(animation.value, [0, 1], [50, 0], Extrapolate.CLAMP),
    opacity: interpolate(animation.value, [0, 1], [0, 1])
  }));

  const animationStyleForm = useAnimatedStyle(() => ({
    opacity: interpolate(animation.value, [0, 1], [0, 1])
  }));

  const {
    control,
    formState: { errors },
    clearErrors,
    getValues,
    setError,
    handleSubmit
  } = useForm<IFormSignIn>({
    resolver: yupResolver(schema),
    mode: 'onSubmit',
    reValidateMode: 'onChange'
  });

  const onSubmit = (form: IFormSignIn) => {
    Keyboard.dismiss();
    dispatch(LOGIN(form));
  };

  useEffect(() => {
    animation.value = withTiming(1, { duration: 1200, easing: Easing.in(Easing.ease) });
  }, []);

  useEffect(() => {
    isObjectEmpty(errors) &&
      Toast.show({
        type: 'error',
        text1: 'Ops... algo não funcionou corretamente',
        text2: errors.email?.message || errors.password?.message,
        props: {
          style: {
            background: 'red'
          }
        }
      });
  }, [errors]);

  useEffect(() => {
    if (isObjectEmpty(error)) {
      setError('email', new Error('Login inválido. Tente novamente'));
      setError('password', new Error('Login inválido. Tente novamente'));
      Toast.show({
        type: 'error',
        text1: 'Ops... algo não funcionou corretamente',
        text2: errors.email?.message || errors.password?.message,
        props: {
          style: {
            background: 'red'
          }
        }
      });
    }
  }, [error]);

  return (
    <TouchableWithoutFeedback testID='SignInScreen' onPress={Keyboard.dismiss}>
      <KeyboardAvoidingView
        style={{ flex: 1 }}
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      >
        <StyledContainer>
          <StatusBar barStyle='light-content' backgroundColor='transparent' translucent />
          <StyledBackground source={background} />
          <StyledTitle>
            <Animated.View style={[animationStyleLogo]}>
              <LogoSvg />
            </Animated.View>
            <Animated.View style={[animationStyleTitle]}>
              <TitleSvg />
            </Animated.View>
          </StyledTitle>
          <AnimatedForm style={[animationStyleForm]}>
            <Input
              name='email'
              placeholder='Email'
              autoCorrect={false}
              autoCapitalize='none'
              control={control}
              error={errors['email']}
              canClearErrorOnFocus={isObjectEmpty(error)}
              getValue={getValues}
              clearErrors={clearErrors}
              errors={errors}
            />
            <Input
              name='password'
              placeholder='Senha'
              secureTextEntry
              autoCorrect={false}
              autoCapitalize='none'
              control={control}
              error={errors['password']}
              getValue={getValues}
              canClearErrorOnFocus={isObjectEmpty(error)}
              onSubmit={handleSubmit(onSubmit)}
              clearErrors={clearErrors}
              errors={errors}
            />
          </AnimatedForm>
        </StyledContainer>
      </KeyboardAvoidingView>
    </TouchableWithoutFeedback>
  );
}

const StyledContainer = styled.View`
  flex: 1;
  justify-content: center;
  padding: 0 ${RFValue(16)}px;
`;

const StyledBackground = styled.ImageBackground`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`;

const StyledTitle = styled.View`
  flex-direction: row;
  align-items: center;
  margin-bottom: ${RFValue(50)}px;
`;

const StyledForm = styled.View``;

const AnimatedForm = Animated.createAnimatedComponent(StyledForm);
