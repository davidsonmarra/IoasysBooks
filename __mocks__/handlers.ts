import { rest } from 'msw';
import { mockFetchDataBooks, mockFetchDataRefresh, mockFetchDataSignIn } from '.';
import constants from '../src/constants';

export const handlers = [
  rest.get(`${constants.BASE_URL}/books`, (req, res, ctx) => {
    const search: string = req.url.searchParams.get('title') || '';
    return search.includes('error')
      ? res(
          ctx.status(401),
          ctx.json({
            errorMessage: `Error 401`
          })
        )
      : res(ctx.json(mockFetchDataBooks));
  }),
  rest.post(`${constants.BASE_URL}/auth/sign-in`, async (req, res, ctx) => {
    const { email } = await req.json();
    return email.includes('error')
      ? res(
          ctx.status(401),
          ctx.json({
            errorMessage: `Error 401`
          })
        )
      : res(ctx.json(mockFetchDataSignIn));
  }),
  rest.post(`${constants.BASE_URL}/auth/refresh-token`, async (req, res, ctx) => {
    const { refreshToken } = await req.json();
    return refreshToken.includes('error')
      ? res(
          ctx.status(401),
          ctx.json({
            errorMessage: `Error 401`
          })
        )
      : res(ctx.json(mockFetchDataRefresh));
  })
];
