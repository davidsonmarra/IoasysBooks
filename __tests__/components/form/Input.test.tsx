import { Input } from '../../../src/components';
import { mockPreloadedStateFilled } from '../../../__mocks__';
import {
  matchSnapshotWithProviders,
  fireEvent,
  renderWithCustomProviders,
  cleanup,
  renderHook
} from '../../test-utils';
import { server } from '../../jest.setup';
import { yupResolver } from '@hookform/resolvers/yup';
import { schema } from '../../../src/presentational/SignIn/schema';
import { FieldError, FieldErrorsImpl, useForm } from 'react-hook-form';
import { IFormSignIn } from '../../../src/presentational/SignIn';

const methods = renderHook(() =>
  useForm<IFormSignIn>({
    resolver: yupResolver(schema),
    mode: 'onSubmit',
    reValidateMode: 'onChange'
  })
);
const mockGetValue = jest.fn(() => 'Harry Potter');

const params = (
  name: keyof IFormSignIn = 'email',
  placeholder = '',
  control = methods.result.current.control,
  error: FieldError | undefined = undefined,
  onSubmit = jest.fn(),
  clearErrors = jest.fn(),
  getValue = mockGetValue,
  errors: Partial<FieldErrorsImpl<IFormSignIn>> = {}
) => ({
  name,
  placeholder,
  control,
  error,
  onSubmit,
  clearErrors,
  getValue,
  errors
});

describe('Form Input', () => {
  afterEach(() => {
    cleanup();
    server.resetHandlers();
    mockGetValue.mockReset();
  });

  afterAll(() => server.close());

  it('should match a snapshot', () => {
    matchSnapshotWithProviders(<Input {...params()} />);
  });

  it('should render input correctly', () => {
    const {
      render: { getByTestId }
    } = renderWithCustomProviders(<Input {...params()} />, mockPreloadedStateFilled);
    const input = getByTestId('input-email');
    fireEvent(input, 'onFocus');
    fireEvent(input, 'onBlur');
    expect(mockGetValue.mock.calls).toEqual([['email'], ['email'], ['email']]);
  });

  it('should clear erros works correctly', () => {
    const mockClearErrors = jest.fn();
    const {
      render: { getByTestId }
    } = renderWithCustomProviders(
      <Input
        {...params(
          'email',
          '',
          methods.result.current.control,
          { type: 'required' },
          jest.fn(),
          mockClearErrors,
          mockGetValue,
          {}
        )}
        canClearErrorOnFocus
      />,
      mockPreloadedStateFilled
    );
    const input = getByTestId('input-email');
    fireEvent(input, 'onFocus');

    expect(mockClearErrors.mock.calls).toEqual([['email']]);
  });
});
