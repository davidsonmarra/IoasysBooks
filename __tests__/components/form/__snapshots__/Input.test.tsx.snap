// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Input should match a snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  forwardedRef={[Function]}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#00000052",
      "borderColor": "rgba(0, 0, 0, 0)",
      "borderRadius": 8,
      "borderWidth": 2,
      "flexDirection": "row",
      "height": 118,
      "justifyContent": "flex-end",
      "marginBottom": 31,
      "opacity": 1,
      "width": "100%",
    }
  }
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    style={
      Array [
        Object {
          "alignSelf": "flex-end",
          "color": "#FFFFFF",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "Heebo-Regular",
          "fontSize": 31,
          "paddingBottom": 16,
          "paddingLeft": 31,
          "paddingRight": 31,
          "paddingTop": 0,
        },
      ]
    }
    testID="input-email"
  />
  <Text
    animatedStyle={
      Object {
        "value": Object {
          "bottom": 38,
          "fontSize": 12,
          "left": 10,
        },
      }
    }
    collapsable={false}
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Heebo-Regular",
          "fontSize": 31,
          "opacity": 0.5,
          "position": "absolute",
        },
        Object {
          "bottom": 38,
          "fontSize": 12,
          "left": 10,
        },
      ]
    }
  />
</View>
`;
